lib/asciidoctor/syntax_highlighter.rb

```diff
@@ -0,0 +1,158 @@
+module Asciidoctor
+  module SyntaxHighlighter
+    # Public: Returns the String name for referencing to this syntax highlighter in messages and option names.
+    attr_reader :name
+
+    def initialize name, doc, opts = {}
+      @name = name
+    end
+
+    # Public: Indicates whether this syntax highlighter has docinfo (i.e., markup) to contribute to the output document
+    # at the specified location.
+    #
+    # Returns a Boolean indicating whether the docinfo method should be called for this location.
+    def docinfo? location; end
+
+    # Public: Generates docinfo markup to include at the specified location in the output document.
+    #
+    # Return a markup String to include in the specified location of the output document.
+    def docinfo location
+      raise ::NotImplementedError, 'Asciidoctor::SyntaxHighlighter must implement the docinfo method if the docinfo? method returns true'
+    end
+
+    # Public: Indicates whether this syntax highlighter highlights during conversion or in the client.
+    #
+    # Returns a Boolean indicating whether the highlight method should be used to handle the :specialchars substitution.
+    def highlight?; end
+
+    # Public: Highlights the specified source when this block is being converted.
+    #
+    # node   - The source block to syntax highlight.
+    # source - The raw source text of this source block (after preprocessing).
+    # lang   - The source language specified on this block (e.g., ruby).
+    # opts   - A Hash of options that control syntax highlighting:
+    #          :callouts - a Hash of callouts that were extracted from the source (index by line number)
+    #          :css_mode - the Symbol CSS mode (class or inline)
+    #          :line_numbers - a Symbol indicating whether line numbers are enabled (nil, :table, or :inline)
+    #          :start_line_number - the starting line number to use when numbering lines
+    #          :highlight_lines - lines in the source block to emphasize (not be confused with syntax highlighting)
+    #          :style - the String style (aka theme) to use
+    #
+    # Returns the highlighted String source.
+    def highlight node, source, lang, opts
+      raise ::NotImplementedError, 'Asciidoctor::SyntaxHighlighter must implement the highlight method if the highlight? method returns true'
+    end
+
+    # Public: Format the highlighted source for inclusion in an HTML document.
+    #
+    # node   - The source block being processed.
+    # lang   - The source language for this block (e.g., ruby).
+    # opts   - A Hash of options that control syntax highlighting:
+    #          :nowrap - a Boolean that indicates that wrapping should be disabled
+    #
+    # Returns the highlighted source String wrapped in preformatted tags (e.g., pre and code)
+    def format node, lang, opts
```

Comment:
@Mogztter Do you think we should provide a more complete default implementation for format? If not in the SyntaxHighlighter module, perhaps in the Base class? The reason I hadn't done it is because it couples us rather tightly to HTML, but then again so do all the implementations so far. I'm thinking the Base class is the right place for it. The idea would be you could skip this method unless you really needed specific HTML...and we could encourage a standard format.

Reply:
I'm going to proceed with putting this default impl in the Base class.